runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -129,6 +129,14 @@ static void* chunk_alloc(void *chunk, size_t size, size_t alignment, bool *zero,
       return NULL;
     }
 
+    // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
+    // chunk on success or NULL on error"
+    if (chunk && chunk != cur_chunk_base) {
```

Comment:
It's hard to imagine that `chunk == cur_chunk_base` is possible here. We know `cur_chunk_base != NULL` because we didn't return above. So `cur_chunk_base` is newly allocated memory, and surely it cannot be the same address as the previously existing chunk, can it? So really, we're just going to return `NULL` here if the passed-in `chunk != NULL`. Plus, if we do return here don't we leak the `cur_chunk_base` we just allocated? It's seems like we'd be better off just returning `NULL` if `chunk != NULL` at the top of the `heap.type == DYNAMIC` block, i.e., we simply can't reallocate chunks in place with a dynamic heap.

Reply:
Yeah, good catch on the leak.

I copied the code from the fixed heap path, but that doesn't have the same problem of leaking memory. When I originally wrote that code I wasn't sure when `chunk` could be non `NULL`. I guessed it might be when jemalloc wanted an allocated at a specific spot (maybe to get a new chunk adjacent to an old one) but it did seem unlikely we would end up doing so.

I think that jemalloc ends up only calling this to grow in place, though you could imagine it might request an adjacent chunk to then merge chunks and grow in place. Anyways, that seems unlikely so I agree we should just return NULL if chunk is non-NULL. 

I was wondering if we should try calling the regMemRealloc API if chunk is non-NULL, but given that chunks are sized based on hugepages we'll never be able to grow-in-place either so that seems like a waste.

Reply:
> the fixed heap path [...] doesn't have the same problem of leaking memory

Right, because at the point where we return we haven't yet updated the fixed-heap management variables. Yeah, I had to read through that to remember how this worked before commenting.

Reply:
I looked into this a little more and I believe that the way jemalloc grows in place is actually to try and allocate new chunk after the old one and if that was successful it will merge the two chunks. So with a fixed heap if I enable merging and do serial allocations the current code actually allows us to realloc in place.

Always returning NULL for dynamic heap extensions will eliminate the possibility of doing large reallocs in place, but I think that is still better than the alternative of trying to allocate and then free'ing if it didn't have the address we want, which will be likely be what happens in the steady state for real codes (not ones that are just serially reallocing in a loop)

Reply:
Pushed a fix for this.